src/main/java/com/iexec/sms/secret/app/TeeTaskRuntimeSecretService.java

```diff
@@ -0,0 +1,117 @@
+    private final EncryptionService encryptionService;
+
+    protected TeeTaskRuntimeSecretService(
+            TeeTaskRuntimeSecretRepository teeTaskRuntimeSecretRepository,
+            EncryptionService encryptionService) {
+        this.teeTaskRuntimeSecretRepository = teeTaskRuntimeSecretRepository;
+        this.encryptionService = encryptionService;
+    }
+
+    /**
+     * Retrieve a secret.
+     * Decrypt if required.
+     */
+    public Optional<TeeTaskRuntimeSecret> getSecret(
+            DeployedObjectType deployedObjectType,
+            String deployedObjectAddress,
+            OwnerRole secretOwnerRole,
+            String owner,
+            long secretIndex,
+            boolean shouldDecryptValue) {
+        deployedObjectAddress = deployedObjectAddress.toLowerCase();
+        final TeeTaskRuntimeSecret wantedSecret = new TeeTaskRuntimeSecret(
+                deployedObjectType,
+                deployedObjectAddress,
+                secretOwnerRole,
+                owner,
+                secretIndex,
+                null
+        );
+        final ExampleMatcher exampleMatcher = ExampleMatcher.matching()
+                .withIgnorePaths("value");
+        final Optional<TeeTaskRuntimeSecret> oSecret = teeTaskRuntimeSecretRepository
+                .findOne(Example.of(wantedSecret, exampleMatcher));
+        if (oSecret.isEmpty()) {
+            return Optional.empty();
+        }
+        if (shouldDecryptValue) {
+            final TeeTaskRuntimeSecret secret = oSecret.get();
+            final String decryptedValue = encryptionService.decrypt(secret.getValue());
+            secret.setValue(decryptedValue);
+        }
+        return oSecret;
+    }
+
+    /**
+     * Check whether a secret exists.
+     *
+     * @return {@code true} if the secret exists in the database, {@code false} otherwise.
+     */
+    public boolean isSecretPresent(DeployedObjectType deployedObjectType,
+                                   String deployedObjectAddress,
+                                   OwnerRole secretOwnerRole,
+                                   String owner,
+                                   long secretIndex) {
+        return getSecret(
+                deployedObjectType,
+                deployedObjectAddress,
+                secretOwnerRole,
+                owner,
+                secretIndex,
+                false
+        ).isPresent();
+    }
+
+    /**
+     * Encrypt secrets and store them.
+     */
+    public void encryptAndSaveSecret(DeployedObjectType deployedObjectType,
+                                     String deployedObjectAddress,
+                                     OwnerRole secretOwnerRole,
+                                     String owner,
+                                     long secretIndex,
+                                     String secretValue) {
+        deployedObjectAddress = deployedObjectAddress.toLowerCase();
```

Comment:
I know we already have a isSecretPresent check in controller but it's probably safer to add a check at this method level, since this service is meant to be "globally" used within the app.
It is fine keeping both checks (controllers + this service level). This part of the app is sensitive so we can keep controller checks for user experience and service check for protecting business logic.

Reply:
Sure, here it is :)